src/rulesets/LinearAlgebra/dense.jl

```diff
@@ -394,3 +394,64 @@ function rrule(
     end
     return Ω, lyap_pullback
 end
+
+#####
+##### `kron`
+#####
+
+@static if VERSION ≥ v"1.9.0"
+    function frule((_, Δx, Δy), ::typeof(kron), x::AbstractVecOrMat{<:Number}, y::AbstractVecOrMat{<:Number})
+        return kron(x, y), kron(Δx, y) + kron(x, Δy)
+    end
+
+    function rrule(::typeof(kron), x::AbstractVector{<:Number}, y::AbstractVector{<:Number})
+        project_x = ProjectTo(x)
+        project_y = ProjectTo(y)
+        function kron_pullback(z̄)
+            dz = reshape(unthunk(z̄), length(y), length(x))
+            x̄ = @thunk(project_x(conj.(dz' * y)))
+            ȳ = @thunk(project_y(dz * conj.(x)))
+            return NoTangent(), x̄, ȳ
+        end
+        return kron(x, y), kron_pullback
+    end
+
+    function rrule(::typeof(kron), x::AbstractMatrix{<:Number}, y::AbstractVector{<:Number})
+        project_x = ProjectTo(x)
+        project_y = ProjectTo(y)
+        function kron_pullback(z̄)
+            dz = reshape(unthunk(z̄), length(y), size(x)...)
+            x̄ = @thunk(project_x(dot.(Ref(y), eachslice(dz; dims = (2, 3)))))
+            ȳ = @thunk(project_y(dot.(Ref(x), eachslice(dz; dims = 1))))
+            return NoTangent(), x̄, ȳ
+        end
+        return kron(x, y), kron_pullback
+    end
+
+    function rrule(::typeof(kron), x::AbstractVector{<:Number}, y::AbstractMatrix{<:Number})
+        project_x = ProjectTo(x)
+        project_y = ProjectTo(y)
+        function kron_pullback(z̄)
+            dz = reshape(unthunk(z̄), size(y, 1), length(x), size(y, 2))
+            x̄ = @thunk(project_x(dot.(Ref(y), eachslice(dz; dims = 2))))
+            ȳ = @thunk(project_y(dot.(Ref(x), eachslice(dz; dims = (1, 3)))))
+            return NoTangent(), x̄, ȳ
+        end
+        return kron(x, y), kron_pullback
+    end
+
+    function rrule(::typeof(kron), x::AbstractMatrix{<:Number}, y::AbstractMatrix{<:Number})
+        project_x = ProjectTo(x)
+        project_y = ProjectTo(y)
+        function kron_pullback(z̄)
+            dz = reshape(unthunk(z̄), size(y, 1), size(x, 1), size(y, 2), size(x, 2))
+            x̄ = @thunk(project_x(_dot_collect.(Ref(y), eachslice(dz; dims = (2, 4)))))
+            ȳ = @thunk(project_y(_dot_collect.(Ref(x), eachslice(dz; dims = (1, 3)))))
```

Comment:
I was wondering if you have to make slices, given that kron is just reshape and `.*`. So here's an attempt to do without:

```julia
using ChainRulesCore

function pr_rule(x::AbstractMatrix{<:Number}, y::AbstractMatrix{<:Number})  # from https://github.com/JuliaDiff/ChainRules.jl/pull/741
    project_x = ProjectTo(x)
    project_y = ProjectTo(y)
    function kron_pullback(z̄)
        dz = reshape(unthunk(z̄), size(y, 1), size(x, 1), size(y, 2), size(x, 2))
        x̄ = @thunk(project_x(dot.(Ref(y), eachslice(dz; dims = (2, 4)))))
        ȳ = @thunk(project_y(dot.(Ref(x), eachslice(dz; dims = (1, 3)))))
        return NoTangent(), x̄, ȳ
    end
end

# using TensorCast
# mykron(x,y) = @cast z[(a,b), (c,d)] := x[b,d] * y[a,c]
# @pretty @cast z[(a,b), (c,d)] := x[b,d] * y[a,c]

function shape_rule(x::AbstractMatrix, y::AbstractMatrix)
    function back(dz)
        x4 = reshape(x, 1, size(x,1), 1, size(x,2))
        y4 = reshape(y, size(y,1), 1, size(y,2), 1)
        dz4 = reshape(unthunk(dz), size(y,1), size(x,1), size(y,2), size(x,2))
        dx = @thunk ProjectTo(x)(reshape(sum(dz4 .* y4, dims=(1,3)), size(x)))  # might be missing conj
        dy = @thunk ProjectTo(y)(reshape(sum(dz4 .* x4, dims=(2,4)), size(y)))
        0, dx, dy
      end
end

let x = rand(10,10), y = rand(10,10)
  b1 = pr_rule(x, y)
  b2 = shape_rule(x, y)
  z = kron(x,y)
  _, dx1, _ = @btime map(unthunk, $b1($z))
  _, dx2, _ = @btime map(unthunk, $b2($z))
  dx1 ≈ dx2
end
#   min 28.250 μs, mean 30.315 μs (4 allocations, 1.88 KiB)
#   min 25.208 μs, mean 43.459 μs (32 allocations, 159.28 KiB)
# true
```
It's a pity to allocate these big arrays `dz4 .* y4` but still seems quicker. Possibly we could use lazy broadcasting to avoid that:
```julia
bc = Broadcast.instantiate(Broadcast.broadcasted(*, [1 2 3], [4, 5]));
sum(bc)  # OK
sum(bc; dims=1) # ERROR: MethodError: no method matching reducedim_init(::typeof(identity), ::typeof(Base.add_sum), ::Base.Broadcast.Broadcasted{…}, ::Int64)
sum!([0 0 0], bc)  # ERROR: MethodError: no method matching sum!(::Matrix{Int64}, ::Base.Broadcast.Broadcasted
sum(bc; dims=1, init=0.0)  # OK, not sure if it's fast or not
```

Reply:
Does this result scale to larger arrays?

Reply:
Result meaning speed difference? It will vary with size & machine. On very small arrays reshaping is faster.

Issues with StaticArrays will be similar at all sizes.

I think broadcasting over slices will work badly on CuArrays, and tend to make Arrays. But right now neither idea seems to work, not sure why 
```
julia> using Metal

julia> bk = pr_rule(MtlArray(rand(Float32, 3,3)), MtlArray(rand(Float32, 3,3)));

julia> bk(MtlArray(rand(Float32, 9,9)))[2] |> unthunk
ERROR: GPU compilation of MethodInstance for (::GPUArrays.var"#broadcast_kernel#26")(::Metal.mtlKernelContext, ::MtlDeviceMatrix{…}, ::Base.Broadcast.Broadcasted{…}, ::Int64) failed
KernelError: passing and using non-bitstype argument

julia> bk2 = shape_rule(MtlArray(rand(Float32, 3,3)), MtlArray(rand(Float32, 3,3)));

julia> bk2(MtlArray(rand(Float32, 9,9)))[2] |> unthunk
ERROR: could not load symbol "LLVMExtraAddPropagateJuliaAddrspaces":
dlsym(RTLD_DEFAULT, LLVMExtraAddPropagateJuliaAddrspaces): symbol not found
```

Reply:
If the `reshape` version is not strictly better than the current one, especially for large arrays, I would propose to keep the current version and put further optimizations in a separate PR.

Reply:
A bit curious at what sizes it's slower for you?

Reply:
I haven't benchmarked anything myself yet. I will give it a go later.

Reply:
One way to ensure any implementation isn't excluding all GPU array types would be to toss a `@gpu` in front of the new tests, no?